src/ImageSharp/Processing/Processors/Dithering/PaletteDitherProcessorBase.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System.Collections.Generic;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Processing.Processors
+{
+    /// <summary>
+    /// The base class for dither and diffusion processors that consume a palette.
+    /// </summary>
+    internal abstract class PaletteDitherProcessorBase<TPixel> : ImageProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        private readonly Dictionary<TPixel, PixelPair<TPixel>> cache = new Dictionary<TPixel, PixelPair<TPixel>>();
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PaletteDitherProcessorBase{TPixel}"/> class.
+        /// </summary>
+        /// <param name="palette">The palette to select substitute colors from.</param>
+        public PaletteDitherProcessorBase(TPixel[] palette)
+        {
+            Guard.NotNull(palette, nameof(palette));
+            this.Palette = palette;
+        }
+
+        /// <summary>
+        /// Gets the palette to select substitute colors from.
+        /// </summary>
+        public TPixel[] Palette { get; }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        protected PixelPair<TPixel> GetClosestPixelPair(ref TPixel pixel, TPixel[] colorPalette)
+        {
+            // Check if the color is in the lookup table
+            if (this.cache.ContainsKey(pixel))
+            {
+                return this.cache[pixel];
+            }
+
+            // Not found - loop through the palette and find the nearest match.
+            float leastDistance = int.MaxValue;
+            float secondLeastDistance = int.MaxValue;
+            var vector = pixel.ToVector4();
+
+            var closest = default(TPixel);
+            var secondClosest = default(TPixel);
+            for (int index = 0; index < colorPalette.Length; index++)
+            {
+                TPixel temp = colorPalette[index];
+                float distance = Vector4.Distance(vector, temp.ToVector4());
```

Comment:
If you are looking for a minimum (and not actually a distance value), it's enough to call Vector4.DistanceSquared() to spare an expensive square root calculation.

Reply:
We're looking for the distance here, basically we want to find the two closest colors to the given one.